File: DS/C++_Programming/Sorting/Bubble_Sort/Program01.cpp
#include<iostream>
using namespace std;

class ArrayX
{
    public:
        int *Arr;
        int iSize;
        bool Sorted;

        // Constructor, Destructor
        ArrayX(int no);
        ~ArrayX();

        // Accept the values and Display the values
        void Accept();
        void Display();

        bool CheckSorted();

        void BubbleSort();
};

ArrayX :: ArrayX(int no)
{
    cout<<"Inside constructor\n";
    iSize = no;
    Arr = new int[iSize];
    Sorted = true;
}

ArrayX :: ~ArrayX()
{
    cout<<"Inside destructor\n";
    delete [] Arr;
}

void ArrayX :: Accept()            // Time Complexity N. It will sort while getting the input
{  
    int iCnt = 0;

    cout<<"Enter the elements : \n";

    cin >> Arr[iCnt];               // Get first input outside the loop

    for(iCnt = 1; iCnt < iSize; iCnt++) // Start loop from 1st index
    {
        cin >> Arr[iCnt];

        if(Arr[iCnt - 1] > Arr[iCnt])
        {
            Sorted = false;
        }
    }

    // Sorted = CheckSorted();
}

void ArrayX :: Display()
{
    int iCnt = 0;

    cout<<"Elemets of the array are : \n";

    for(iCnt = 0; iCnt < iSize; iCnt++)
    {
        cout<<Arr[iCnt]<<"\t";
    }

    cout<<"\n";
}

// 0    1   2   3   4   5   6   7   8   9
// 12   14  17  24  28  36  48  52  52  65
bool ArrayX :: CheckSorted()
{
    int i = 0;
    bool bFlag = true;

    for (i = 0; i < iSize - 1; i++)
    {
        if(Arr[i] > Arr[i + 1])
        {
            bFlag = false;
            break;
        }
    }

    return bFlag;
}

void ArrayX :: BubbleSort()
{
    int i = 0, j = 0, temp = 0;

    for (i = 0; i < iSize - 1; i++)             // Outer loop
    {
        for (j = 0; j < iSize - 1 - i; j++)     // Inner Loop
        {
            if(Arr[j] > Arr[j + 1])
            {
                temp = Arr[j];
                Arr[j] = Arr[j + 1];
                Arr[j + 1] = temp;
            }
        }
    }

    Sorted = true;
}

int main()
{
    int iValue = 0;

    cout<<"Enter the number of elements : \n";
    cin>>iValue;

    ArrayX aobj(iValue);

    aobj.Accept();

    cout << "\nData Before Sorting :-\n";
    aobj.Display();

    if(aobj.Sorted == true)
    {
        cout << "\nData is sorted.\n";
    }
    else
    {
        cout << "\nData is not sorted.\n";
    }

    aobj.BubbleSort();

    cout << "\nData After Sorting :-\n";
    aobj.Display();

    return 0;
}

File: DS/Java_Programming/Sorting/Bubble_Sort/Program01.java
import java.util.Scanner;

class ArrayX
{
    public int Arr[];
    public int iSize;
    public boolean Sorted;

    // Constructor
    public ArrayX(int no)
    {
        System.out.println("Inside constructor");
        iSize = no;
        Arr = new int[iSize];
        Sorted = true;
    }

    // Accept elements (checks sorted while accepting)
    public void Accept()
    {
        Scanner sobj = new Scanner(System.in);

        System.out.println("Enter the elements : ");

        Arr[0] = sobj.nextInt();   // First element

        for(int iCnt = 1; iCnt < iSize; iCnt++)
        {
            Arr[iCnt] = sobj.nextInt();

            if(Arr[iCnt - 1] > Arr[iCnt])
            {
                Sorted = false;
            }
        }
    }

    // Display elements
    public void Display()
    {
        System.out.println("Elements of the array are : ");

        for(int iCnt = 0; iCnt < iSize; iCnt++)
        {
            System.out.print(Arr[iCnt] + "\t");
        }

        System.out.println();
    }

    // Check if array is sorted
    public boolean CheckSorted()
    {
        boolean bFlag = true;

        for(int i = 0; i < iSize - 1; i++)
        {
            if(Arr[i] > Arr[i + 1])
            {
                bFlag = false;
                break;
            }
        }

        return bFlag;
    }

    // Bubble Sort
    public void BubbleSort()
    {
        int temp = 0;

        for(int i = 0; i < iSize - 1; i++)
        {
            for(int j = 0; j < iSize - 1 - i; j++)
            {
                if(Arr[j] > Arr[j + 1])
                {
                    temp = Arr[j];
                    Arr[j] = Arr[j + 1];
                    Arr[j + 1] = temp;
                }
            }
        }

        Sorted = true;
    }
}

public class Program01
{
    public static void main(String[] args)
    {
        Scanner sobj = new Scanner(System.in);

        int iValue = 0;

        System.out.println("Enter the number of elements : ");
        iValue = sobj.nextInt();

        ArrayX aobj = new ArrayX(iValue);

        aobj.Accept();

        System.out.println("\nData Before Sorting :-");
        aobj.Display();

        if(aobj.Sorted == true)
        {
            System.out.println("\nData is sorted.");
        }
        else
        {
            System.out.println("\nData is not sorted.");
        }

        aobj.BubbleSort();

        System.out.println("\nData After Sorting :-");
        aobj.Display();
    }
}